### PR TITLE
[Test][Softmax] Add non-aligned edge-case shapes for softmax-family ops

### LINKED ISSUE #906: [TEST][Kernel] Add non-aligned M×N edge-case tests for softmax-family kernels
### Symptom / Motivation

Discovered during PR #898 ([Perf][Kernel] Optimise large-N tiled softmax to match PyTorch).

The kernel-side masked-load rewrite initially missed M-dimension row-tail guards (`pid_m * block_m + i < M`), only guarding the N-dimension (`j < N`). This correctness regression was caught by human review (Ibuki-wind), not by the existing test suite. The suite covers non-aligned M with aligned N `(33, 256)` and aligned M with non-aligned N `(32, 300)`, but not the critical combination where **both** M and N are non-aligned simultaneously.

### Root Cause Analysis

The test parametrization doesn't include shapes that exercise both the M-guard and N-guard paths at the same time. Shapes like `(33, 300)`, `(33, 33000)`, and `(33, 32768)` would have caught the regression automatically.

### Related Files

- `tests/ops/test_softmax.py` — PARAMS lists for softmax, log_softmax, logsumexp test classes

## Goal

Add targeted edge-case test shapes that exercise the M×N non-aligned combination for softmax-family kernels, covering single-tile and multi-tile (large-N) paths.

## Plan

1. Add `(33, 300)` shapes (both non-aligned, single-tile path) for softmax, log_softmax, logsumexp
2. Add `(33, 33000)` shapes (both non-aligned, multi-tile path with masked loads) for all three ops
3. Add `(33, 32768)` shapes (non-aligned M, large-N tiled path) for all three ops
4. Verify all new shapes pass on current HEAD

## Constraints

- Must not regress PR #898
- New test shapes should use `pytest.mark.full` tier (not smoke)

## Acceptance Criteria

- [ ] AC-1: Test shapes `(33, 300)`, `(33, 33000)`, `(33, 32768)` added for softmax, log_softmax, and logsumexp
- [ ] AC-2: All new and existing softmax-family tests pass

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Add targeted edge-case test shapes that exercise the M×N non-aligned combination for softmax-family kernels, covering single-tile and multi-tile (large-N) paths.

Closes #906

## Changes

**Tests** (`tests/ops/test_softmax.py`):
- 3 new shapes per fixture (SoftmaxFixture, LogSoftmaxFixture, LogSumExpFixture):
  - `(33, 300)` fp32 — both M and N non-aligned, single-tile path
  - `(33, 33000)` fp16 — both M and N non-aligned, multi-tile with masked loads
  - `(33, 32768)` fp16 — non-aligned M with large-N tiled path

**Scripts** (`scripts/test_node_delta.py`):
- Fix base-ref collection: write temp file in original directory instead of `/tmp/` so pytest can resolve project imports

**Foundry** (`.foundry/mold/pre-create-checks.md`):
- Add test node delta as pre-create check #3 (soft gate) with correct `--base upstream/main` command

## Test plan

- [x] AC-1: All 126 tests pass (`pytest tests/ops/test_softmax.py`)
- [x] AC-2: New shapes exercise non-aligned M×N paths (single-tile and multi-tile)

## Test node delta

```
File                         Base    HEAD    Delta
--------------------------------------------------
tests/ops/test_softmax.py     117     126       +9
--------------------------------------------------
TOTAL                         117     126       +9

Growth: +7.7%
```

Justification: 9 new `pytest.mark.full` nodes (3 shapes × 3 fixtures). Each shape targets a distinct non-aligned code path not covered by existing aligned shapes. No cross-dtype expansion (existing fixtures already cover fp32/fp16/bf16 on aligned shapes; the non-aligned path shares the same tiling/masked-load logic across fp16 and bf16).

## Follow-up

No follow-up issues or suggestions.

### CODE REVIEW

File: tests/ops/test_softmax.py [hunk 1]
```diff
@@ -74,6 +74,12 @@ class SoftmaxFixture(FixtureBase):
                 # dim=-1, large-N (triggers N-tiling path)
                 pytest.param((4, 32768), -1, torch.float16, False, marks=pytest.mark.full),
                 pytest.param((4, 32768), -1, torch.bfloat16, False, marks=pytest.mark.full),
+                # dim=-1, M×N both non-aligned (single-tile path)
+                pytest.param((33, 300), -1, torch.float32, False, marks=pytest.mark.full),
+                # dim=-1, M×N both non-aligned (multi-tile, masked loads)
+                pytest.param((33, 33000), -1, torch.float16, False, marks=pytest.mark.full),
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:80]: Reviewer determined this is a false positive:

tests/ops/test_softmax.py adds (33,33000) as shape-path coverage for the non-aligned multi-tile masked-load branch, and docs/testing.md plus .claude/domain-rules/testing-budget.md explicitly say not to cross dtype and shape unless the combination exercises a distinct code path. Existing fixtures already cover bfloat16 on the large-N tiled path ((4,32768) in all three op families), and the softmax/logsumexp kernels compute tiling from dtype element_size(), so float16 and bfloat16 share the same tiling/masked-load branch here.

Current HEAD: 6d847b6

File: tests/ops/test_softmax.py [hunk 2]
```diff
@@ -212,6 +218,12 @@ class LogSoftmaxFixture(FixtureBase):
...
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:224]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

File: tests/ops/test_softmax.py [hunk 3]
```diff
@@ -277,6 +289,12 @@ class LogSumExpFixture(FixtureBase):
...
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:295]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

File: tests/ops/test_softmax.py [hunk 4]
```diff
@@ -74,6 +74,12 @@ class SoftmaxFixture(FixtureBase):
...
+                # dim=-1, non-aligned M + large-N tiled path
+                pytest.param((33, 32768), -1, torch.float16, False, marks=pytest.mark.full),
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:82]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:82]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

File: tests/ops/test_softmax.py [hunk 5]
```diff
@@ -212,6 +218,12 @@ class LogSoftmaxFixture(FixtureBase):
...
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:226]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

File: tests/ops/test_softmax.py [hunk 6]
```diff
@@ -277,6 +289,12 @@ class LogSumExpFixture(FixtureBase):
...
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:297]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

File: tests/ops/test_softmax.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_softmax.py:82]: Not applicable — bfloat16 and float16 share the same tiling and masked-load code path in the softmax kernel (tiling is computed from `dtype.element_size()`, which is 2 for both). Existing fixtures already cover bfloat16 on the large-N tiled path via `(4, 32768)` in all three op families. Per project testing guidelines, dtype x shape cross-products are not added unless the combination exercises a distinct code path.

Current HEAD: 6d847b6

File: .foundry/mold/benchmark-template.md [hunk 7]
```diff
@@ -17,7 +17,7 @@
 | `"tileops"`  | TileOPs implementation (required, exactly once per config) |
 | `"torch"`    | PyTorch reference implementation                           |
 | `"FA3"`      | FlashAttention 3                                           |
-| `"fla"`      | Functional Linear Attention                                |
+| `"FLA"`      | Flash Linear Attention                                     |
```
Comment (Ibuki-wind) [inline .foundry/mold/benchmark-template.md]: The benchmark-template tag convention is changing from the repo's current / usage to /, and this file is outside the  scope of the PR. That will make new benchmarks diverge from the existing docs/files and mixes an unrelated benchmark-contract change into a test-only PR -> remove  from this PR, or move a repo-wide tag/docs migration into a dedicated benchmark/docs PR.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Test-stage changes in  stay within the trust boundary, and the new shapes are targeted at the non-aligned MxN paths.\n\nBlocking issue: this PR also edits , which is outside a  change and changes the benchmark tag convention away from the repo's current / usage without a coordinated benchmark/docs migration.\n\nPlease drop or split the benchmark-template edit, or update the benchmark/docs/reporting contract consistently in a dedicated PR.\n\nCross-cutting process issue: this PR grows an existing test file, so the PR body should include the No changed test files detected. output plus a one-line justification per .

File: .foundry/mold/benchmark-template.md (hunk 7 above)
Comment (Ibuki-wind) [inline .foundry/mold/benchmark-template.md]: The benchmark-template tag convention is changing from the repo's current `fa3`/`fla` usage to `FA3`/`FLA`, and this file is outside the `[Test][Softmax]` scope of the PR. That will make new benchmarks diverge from the existing docs/files and mixes an unrelated benchmark-contract change into a test-only PR -> remove `.foundry/mold/benchmark-template.md` from this PR, or move a repo-wide tag/docs migration into a dedicated benchmark/docs PR.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Test-stage changes in `tests/ops/test_softmax.py` stay within the trust boundary, and the new shapes are targeted at the non-aligned MxN paths.

Blocking issue: this PR also edits `.foundry/mold/benchmark-template.md`, which is outside a `[Test][Softmax]` change and changes the benchmark tag convention away from the repo's current `fa3`/`fla` usage without a coordinated benchmark/docs migration.

Please drop or split the benchmark-template edit, or update the benchmark/docs/reporting contract consistently in a dedicated PR.

Cross-cutting process issue: this PR grows an existing test file, so the PR body should include the `scripts/test_node_delta.py` output plus a one-line justification per `testing-budget.md`.

File: .foundry/mold/benchmark-template.md (hunk 7 above)
Comment (lcy-seso) [reply .foundry/mold/benchmark-template.md]: Adopted. Reverted the benchmark-template change from this PR. See aa16811.

Comment (lcy-seso) [reply .foundry/mold/benchmark-template.md]: Adopted. Reverted the benchmark-template change from this PR. See aa16811.

Comment (lcy-seso) [reply .foundry/mold/benchmark-template.md]: Overridden by human developer. The FLA tag fix is intentionally bundled here. Restored in 291962c.

Comment (lcy-seso) [reply .foundry/mold/benchmark-template.md]: Overridden by human developer. The FLA tag fix is intentionally bundled here. Restored in 291962c.

Comment (Ibuki-wind) [inline .foundry/mold/benchmark-template.md]: The benchmark-template tag convention here conflicts with the benchmark rule, which explicitly requires lowercase tags such as `fa3` and `fla`, and this benchmark-contract edit is outside the `[Test][Softmax]` scope of the PR → remove `.foundry/mold/benchmark-template.md` from this PR, or move the tag/docs migration into a dedicated benchmark/docs PR that updates the contract consistently.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Trust-boundary verdict: this should still be split before merge. The softmax-family test additions are targeted and consistent with the testing rules, but the PR also carries a benchmark-template contract change in the same `[Test][Softmax]` review.

Cross-cutting process issue: because `tests/ops/test_softmax.py` grows an existing test file, the PR body should include `scripts/test_node_delta.py` output plus a one-line justification per `testing-budget.md`.

Main risk is not the new test shapes; it is broadening the PR into benchmark-contract churn while the functional test change itself is otherwise clean.

File: .foundry/mold/benchmark-template.md (hunk 7 above)
Comment (lcy-seso) [reply .foundry/mold/benchmark-template.md]: Adopted. Benchmark domain rule (`.claude/domain-rules/benchmark.md:13`) requires lowercase tags (`fa3`, `fla`). Template change reverted. See 7b4ac56.

**Review by Ibuki-wind — COMMENTED** [review]
Trust-boundary verdict: clean. The PR now stays entirely in `tests/`, and the added `(33, 300)`, `(33, 33000)`, and `(33, 32768)` cases are targeted at distinct non-aligned / large-N softmax-family paths rather than combinatorial coverage.

Only remaining note: because `tests/ops/test_softmax.py` grows an existing test file, the PR body should include the `scripts/test_node_delta.py` output plus a one-line justification per `testing-budget.md`.

**Review by Ibuki-wind — APPROVED** [review]
Trust-boundary verdict: clean. The PR now stays within test-stage/supporting test-tooling concerns, and the `test_node_delta.py` fix directly supports the required node-growth reporting for this test expansion.

The added softmax-family shapes are targeted at distinct non-aligned / large-N paths, and the updated node-delta output is consistent with the helper script when run on the PR head.

File: .foundry/mold/pre-create-checks.md [hunk 8]
```diff
@@ -25,3 +25,28 @@ If PR introduces or modifies a kernel/op:
...
+python scripts/test_node_delta.py --base upstream/main tests/ops/test_<name>.py
+```
+
+**Interpreting output:**
+
+- **No growth on existing files** → nothing to report.
+- **Growth on existing files** → include script output and one-line justification in PR body under `## Test node delta`.
+- **New test files only** (delta shows "all N nodes from new files") → no delta report needed.
+
+**SOFT GATE:** Does not block PR creation, but missing justification will be flagged during review (per `testing-budget.md`).
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply .foundry/mold/pre-create-checks.md]: Adopted. Fixed path to `.claude/domain-rules/testing-budget.md`. See e5e5dd4.
